### PR TITLE
perf(qwen3.6): fused router GEMV + bitonic top-k (grid-completion, decode)

### PR DESCRIPTION
**Pull request by fansilas**
## Summary

One decode critical-path latency cut for Qwen3.6-35B-A3B (256-expert MoE), bit-identical to main's output.

**Fused router GEMV + top-k.** The router today runs a split-K GEMV that writes 256 expert logits to global memory, then a *separate* single-block kernel reads them back and selects the top-8 with an O(num_experts^2) rank scan (`moe_router_kernel2`). On an nsys node trace that top-k kernel is ~4.6% of the 128-context token, sitting serial on the decode critical path (router -> gate/up).

This PR fuses both into one kernel (`moe_router_fused_kernel`): the same split-K GEMV writes its logits, then a grid-completion handshake (an `atomicInc` that wraps back to 0, so it is CUDA-graph-replay safe with no reset node) lets the last block run an in-kernel top-8.

The select is a bitonic reduction tree that replaces the 8 serial warp-argmax passes with a 5-level tree of independent (ILP) shuffles, and the fusion removes the separate top-k launch plus its dependent-load gap. Selection is byte-identical to the existing rank-select (value descending, ties to the lower expert index; verified by an identical greedy decode over 48 tokens). Only the 256-expert path is fused; the Qwen3-30B guard (128 experts) keeps the existing rank-select unchanged.

Toggle for the A/B below (default ON, 256-expert path): `SPARKINFER_ROUTER_FUSED=0` restores the separate GEMV + top-k.

## Proof of speedup

- [x] Tested on **RTX 5090** (`sm_120`)

**Decode tok/s** (end-to-end, `qwen3_gguf_bench`, `Qwen3.6-35B-A3B-UD-Q4_K_M.gguf`, 128 decode tokens, 128-context):

| | decode tok/s |
|---|--:|
| before (main) | 403.5 |
| after (this PR) | 419.7 |

Interleaved same-binary A/B (median of 4, env toggle flips the fused router), all three scored contexts:

```text
# ctx=128:  before 403.5 -> after 419.7  (+4.0%)
# ctx=512:  before 397.9 -> after 414.4  (+4.2%)
# ctx=4096: before 382.0 -> after 396.9  (+3.9%)

-- ctx=128 (before / after) --
403.44 419.81
403.57 419.68
403.70 419.75
-- ctx=512 --
397.83 414.40
397.85 414.41
397.89 414.50
-- ctx=4096 --
382.31 396.98
381.89 396.93
382.02 396.87
```

Accuracy: the fused router's selection is byte-identical to main, so the decode is unchanged — greedy output matches the baseline exactly (48/48 tokens). The Qwen3-30B-A3B guard path (128-expert router) is untouched.

## Relation to open PRs

`moe_router_fused_kernel` is original and implements an optimization no open PR does. It reuses the base-repo `gemv_f32_sk<float,4>` body (base code, unchanged by any PR) and adds the grid-completion handshake + in-kernel bitonic top-8. The only files shared with open PRs are `moe/router.cu` and `runtime/src/models/qwen35.cpp`:
- vs **#237** (`router.cu`): #237 only tunes the router-GEMV split factor and the `launch_moe_router` dispatch (keeps the warp top-k); it adds no fused kernel and does not rewrite the top-k. This PR does not touch that function and instead adds new symbols. (Added-line containment vs #237 is ~6%.)
- `qwen35.cpp`: the router wiring here (a new `use_router_fused` dispatch + `SPARKINFER_ROUTER_FUSED` env) sits in a disjoint region from other decode PRs' edits.

No residual/norm change is included (main already folds the routed+shared residual via the merged addnorm3).

### CODE REVIEW

**Comment by skyrocket2026** [comment]
## ✅ sparkinfer auto-eval — `bb7a2f5`

| metric | value |
|---|---|
| **label** | `eval:XL` |
| scored decode (128 ctx · 128-context · Qwen3.6) | 427.05 tok/s |
| vs same-box main | 412.53 tok/s → +3.5% (+14.5) |
| correctness (Qwen3.6 vs llama.cpp) | top-1 95.8% · KL 0.019 |
| Qwen3.6 128-token no-regression gate | 427.05 tok/s vs main 412.53 tok/s · pass |
| Qwen3.6 512-context no-regression gate | 419.23 tok/s vs main 405.25 tok/s · pass |
| Qwen3.6 4k-context no-regression gate | 401.55 tok/s vs main 388.62 tok/s · pass |
| Qwen3.6 16k-context no-regression gate | 367.35 tok/s vs main 355.78 tok/s · pass |
| Qwen3.6 32k-context no-regression gate | 327.16 tok/s vs main 319.97 tok/s · pass |
| **Qwen3-30B-A3B guard — accuracy** | top-1 95.8% · KL 0.0161 · pass |
| Qwen3-30B-A3B guard — 128-token | 485.61 tok/s · pass |
| Qwen3-30B-A3B guard — 512-context | 464.1 tok/s · pass |
| Qwen3-30B-A3B guard — 4k-context | 385.58 tok/s · pass |
| Qwen3-30B-A3B guard — 16k-context | 322.0 tok/s · pass |
| Qwen3-30B-A3B guard — 32k-context | 256.93 tok/s · pass |

Verified speedup over same-box origin/main — 427.05 tok/s (main was 412.53 tok/s).

_RTX 5090 (sm_120) · 128/512/4k/16k/32k guarded · scored vs same-box main · strongest context scores · built from source · correctness vs llama.cpp. Automated — **not merged**; merge manually after review._

**Comment by fansilas** [comment]
Hi @skyrocket2026 
I rebased onto current main. Could you please review again? Thanks!

**Comment by skyrocket2026** [comment]
## ✅ sparkinfer auto-eval — `1e75ffe`

| metric | value |
|---|---|
| **label** | `eval:XL` |
| scored decode (128 ctx · 128-context · Qwen3.6) | 426.0 tok/s |
| vs same-box main | 408.95 tok/s → +4.2% (+17.1) |
| correctness (Qwen3.6 vs llama.cpp) | top-1 95.3% · KL 0.0122 |
| Qwen3.6 128-token no-regression gate | 426.0 tok/s vs main 408.95 tok/s · pass |
| Qwen3.6 512-context no-regression gate | 419.18 tok/s vs main 402.91 tok/s · pass |
| Qwen3.6 4k-context no-regression gate | 402.37 tok/s vs main 387.63 tok/s · pass |
| Qwen3.6 16k-context no-regression gate | 385.24 tok/s vs main 372.39 tok/s · pass |
| Qwen3.6 32k-context no-regression gate | 363.48 tok/s vs main 351.42 tok/s · pass |
| **Qwen3-30B-A3B guard — accuracy** | top-1 95.8% · KL 0.0175 · pass |
| Qwen3-30B-A3B guard — 128-token | 490.53 tok/s · pass |
| Qwen3-30B-A3B guard — 512-context | 470.39 tok/s · pass |
| Qwen3-30B-A3B guard — 4k-context | 387.13 tok/s · pass |
| Qwen3-30B-A3B guard — 16k-context | 328.27 tok/s · pass |
| Qwen3-30B-A3B guard — 32k-context | 261.96 tok/s · pass |

Verified speedup over same-box origin/main — 426.0 tok/s (main was 408.95 tok/s).

_RTX 5090 (sm_120) · 128/512/4k/16k/32k guarded · scored vs same-box main · strongest context scores · built from source · correctness vs llama.cpp. Automated — **not merged**; merge manually after review._

**Comment by skyrocket2026** [comment]
✅ Auto-merged as the round's `merge-first` winner — verified same-box speedup over `main`, all checks green. Thanks for the contribution!
